vortex-serde/src/io/dispatcher/compio.rs

```diff
@@ -0,0 +1,100 @@
+use std::future::Future;
+use std::panic::resume_unwind;
+use std::thread::JoinHandle;
+
+use compio::runtime::{JoinHandle as CompioJoinHandle, Runtime, RuntimeBuilder};
+use futures_channel::oneshot;
+use vortex_error::{vortex_bail, vortex_panic, VortexResult};
+
+use super::Dispatch;
+
+trait CompioSpawn {
+    fn spawn(self: Box<Self>) -> CompioJoinHandle<()>;
+}
+
+struct CompioTask<F, R> {
+    task: F,
+    result: oneshot::Sender<R>,
+}
+
+impl<F, Fut, R> CompioSpawn for CompioTask<F, R>
+where
+    F: FnOnce() -> Fut + Send + 'static,
+    Fut: Future<Output = R>,
+    R: Send + 'static,
+{
+    fn spawn(self: Box<Self>) -> CompioJoinHandle<()> {
+        let CompioTask { task, result } = *self;
+        Runtime::with_current(|rt| {
+            rt.spawn(async move {
+                let task_output = task().await;
+                result.send(task_output).ok();
+            })
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct CompioDispatcher {
+    submitter: flume::Sender<Box<dyn CompioSpawn + Send>>,
+    threads: Vec<JoinHandle<()>>,
+}
+
+impl CompioDispatcher {
+    pub fn new(num_threads: usize) -> Self {
+        let (submitter, rx) = flume::unbounded();
+        let threads: Vec<_> = (0..num_threads)
+            .map(|tid| {
+                let worker_thread = std::thread::Builder::new();
+                let worker_thread = worker_thread.name(format!("compio-dispatch-{tid}"));
+                let rx: flume::Receiver<Box<dyn CompioSpawn + Send>> = rx.clone();
+
+                worker_thread
+                    .spawn(move || {
+                        // Create a runtime-per-thread
+                        let rt = RuntimeBuilder::new().build().unwrap_or_else(|e| {
+                            vortex_panic!("CompioDispatcher RuntimeBuilder build(): {e}")
+                        });
```

Comment:
i'm not sure what we can do except panic in this case, unless we make the constructor fallible. but i don't really think this should happen in normal operation, and there's not an obvious way to recover if we cannot build an async runtime